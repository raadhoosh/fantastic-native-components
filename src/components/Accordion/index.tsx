import * as React from "react";
import { StyleSheet, Text, View, ViewStyle } from "react-native";
import { Icon } from "..";
import { Panel } from "..";
type IStyle = ViewStyle | object | Array<ViewStyle>;
interface ObInterface {
    title: string;
    content: string;
}
type Props = {
    dataArray: Array<ObInterface>;
    expanded: number;
    icon?: JSX.Element;
    expandedIcon?: JSX.Element;
    header?: JSX.Element | JSX.Element[];
    body?: JSX.Element | JSX.Element[];
    styledContent?: IStyle;
    primary?: boolean;
    secondary?: boolean;
    success?: boolean;
    info?: boolean;
    warning?: boolean;
    danger?: boolean;
    light?: boolean;
    dark?: boolean;
    backgroundColor?: string;
};
const styles = StyleSheet.create({
    container: {
        backgroundColor: "#fff",
        margin: 10,
        overflow: "hidden",
    },
    titleContainer: {
        flexDirection: "row",
    },
    title: {
        flex: 1,
        padding: 10,
        color: "#2a2f43",
        fontWeight: "bold",
    },
    button: {

    },
    buttonImage: {
        width: 30,
        height: 25,
    },
    body: {
        padding: 10,
        paddingTop: 0,
    },
});
interface State {
    visible: number;
}
class Accordion extends React.Component<Props, State> {
    icons: {
        "up": JSX.Element; "down": JSX.Element;
    };

    constructor(props: Props) {
        super(props);
        this.state = {
            visible: props.expanded,
        };
        this.icons = {
            "up": <Icon color="#fff" type="FontAwesome" name="sort-up" />,
            "down": <Icon color="#fff" type="FontAwesome" name="sort-down" />,
        };
    }
    toggle = (index: number) => () => {
        this.setState({ visible: index });
    }
    renderIcon = (key: number) => {
        const { icon } = this.props;
        let elmIcon = icon || this.icons.down;
        if (this.state.visible === key) {
            elmIcon = this.icons.up;
        }
        return elmIcon;
    }
    renderBody = (item: ObInterface, key: number) => {
        if (this.state.visible === key) {
            return <View>
                <Text>{item.content}</Text>
            </View>;
        }
        return <Text>{""}</Text>;
    }

    render() {
        const { dataArray, icon, expandedIcon, expanded, ...others } = this.props;
        return (<React.Fragment  >
            {dataArray.map((item, index) => (
                <Panel
                    {...others}
                    icon={this.renderIcon(index)}
                    key={index}
                    visible={this.state.visible === index}
                    title={item.title}
                    toggle={this.toggle(index)}
                >
                    {this.renderBody(item, index)}

                </Panel>
            ))}

        </React.Fragment>);
    }
}

export default Accordion;