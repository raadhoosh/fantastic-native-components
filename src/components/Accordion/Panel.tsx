import * as React from "react";
import { StyledHeader, StyledHeaderText, StyledContent, StyledBody } from "./Panel.style";
import { ViewStyle, ModalProps, View, Modal, TouchableOpacity } from "react-native";
import { Icon, Button } from "../../components";
import { Theme } from "..";
type IStyle=  ViewStyle | object | Array<ViewStyle>;
interface IProps {
    header?: JSX.Element | JSX.Element[];
    body?: JSX.Element | JSX.Element[];
    styledContent?: IStyle;
    children?: JSX.Element | JSX.Element[] | string;
    icon?: JSX.Element;
    visible?: boolean;
    primary?: boolean;
    secondary?: boolean;
    success?: boolean;
    info?: boolean;
    warning?: boolean;
    danger?: boolean;
    light?: boolean;
    dark?: boolean;
    backgroundColor?: string;
    title?: string;
    toggle?: () => void;
}

class Panel extends React.PureComponent<IProps> {
    renderHeader = () => {
        const { header } = this.props;
        if (header) {
            return header;
        }
        return (<StyledHeader onPress={this.props.toggle} {...this.props}>
            <StyledHeaderText>
                {this.props.title}
            </StyledHeaderText>
            {this.props.icon}
        </StyledHeader >);
    }
    renderBody = () => {
        const { visible, body } = this.props;
        if (visible) {
            if (body) {
                return body;
            }
            return ((<StyledBody {...this.props}>
                {this.props.children}
            </StyledBody>));
        }
        return null;
    }
    render() {
        const { styledContent } = this.props;
        let style: IStyle = { flex: 1, marginTop: 5 };
        if (styledContent) {
            style = styledContent;
        }
        return (
            <StyledContent style={style}>
                {this.renderHeader()}
                {this.renderBody()}
            </StyledContent>
        );
    }

}

export default Panel;
