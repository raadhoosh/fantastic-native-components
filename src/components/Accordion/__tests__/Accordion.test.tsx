import Accordion from "..";
import { shallow, ShallowWrapper } from "enzyme";
import React from "react";

describe("Accordion", () => {
    describe("rendering", () => {
        let wrapper: ShallowWrapper;
        const dataArray = [
            { title: "First Element", content: "Lorem ipsum dolor sit amet" },
            { title: "Second Element", content: "Lorem ipsum dolor sit amet" },
            { title: "Third Element", content: "Lorem ipsum dolor sit amet" },
          ];
        beforeEach(() => {
            wrapper = shallow(<Accordion primary  dataArray={dataArray} expanded={0} />);
        });
        it("renders", () => {
            expect(wrapper.exists()).toBe(true);
        });
        it("toMatchSnapshot", () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});