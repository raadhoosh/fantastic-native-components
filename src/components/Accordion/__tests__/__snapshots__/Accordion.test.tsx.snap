// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion rendering toMatchSnapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Accordion
    dataArray={
      Array [
        Object {
          "content": "Lorem ipsum dolor sit amet",
          "title": "First Element",
        },
        Object {
          "content": "Lorem ipsum dolor sit amet",
          "title": "Second Element",
        },
        Object {
          "content": "Lorem ipsum dolor sit amet",
          "title": "Third Element",
        },
      ]
    }
    expanded={0}
    primary={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Panel
          icon={
            <ForwardRef
              color="#fff"
              name="sort-up"
              type="FontAwesome"
            />
          }
          primary={true}
          title="First Element"
          toggle={[Function]}
          visible={true}
        >
          <Component>
            <Component>
              Lorem ipsum dolor sit amet
            </Component>
          </Component>
        </Panel>,
        <Panel
          icon={
            <ForwardRef
              color="#fff"
              name="sort-down"
              type="FontAwesome"
            />
          }
          primary={true}
          title="Second Element"
          toggle={[Function]}
          visible={false}
        >
          <Component>
            
          </Component>
        </Panel>,
        <Panel
          icon={
            <ForwardRef
              color="#fff"
              name="sort-down"
              type="FontAwesome"
            />
          }
          primary={true}
          title="Third Element"
          toggle={[Function]}
          visible={false}
        >
          <Component>
            
          </Component>
        </Panel>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "children": <Component>
            <Component>
              Lorem ipsum dolor sit amet
            </Component>
          </Component>,
          "icon": <ForwardRef
            color="#fff"
            name="sort-up"
            type="FontAwesome"
          />,
          "primary": true,
          "title": "First Element",
          "toggle": [Function],
          "visible": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Component>
              Lorem ipsum dolor sit amet
            </Component>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Lorem ipsum dolor sit amet",
            },
            "ref": null,
            "rendered": "Lorem ipsum dolor sit amet",
            "type": [Function],
          },
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "children": <Component>
            
          </Component>,
          "icon": <ForwardRef
            color="#fff"
            name="sort-down"
            type="FontAwesome"
          />,
          "primary": true,
          "title": "Second Element",
          "toggle": [Function],
          "visible": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "children": <Component>
            
          </Component>,
          "icon": <ForwardRef
            color="#fff"
            name="sort-down"
            type="FontAwesome"
          />,
          "primary": true,
          "title": "Third Element",
          "toggle": [Function],
          "visible": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Panel
            icon={
              <ForwardRef
                color="#fff"
                name="sort-up"
                type="FontAwesome"
              />
            }
            primary={true}
            title="First Element"
            toggle={[Function]}
            visible={true}
          >
            <Component>
              <Component>
                Lorem ipsum dolor sit amet
              </Component>
            </Component>
          </Panel>,
          <Panel
            icon={
              <ForwardRef
                color="#fff"
                name="sort-down"
                type="FontAwesome"
              />
            }
            primary={true}
            title="Second Element"
            toggle={[Function]}
            visible={false}
          >
            <Component>
              
            </Component>
          </Panel>,
          <Panel
            icon={
              <ForwardRef
                color="#fff"
                name="sort-down"
                type="FontAwesome"
              />
            }
            primary={true}
            title="Third Element"
            toggle={[Function]}
            visible={false}
          >
            <Component>
              
            </Component>
          </Panel>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "children": <Component>
              <Component>
                Lorem ipsum dolor sit amet
              </Component>
            </Component>,
            "icon": <ForwardRef
              color="#fff"
              name="sort-up"
              type="FontAwesome"
            />,
            "primary": true,
            "title": "First Element",
            "toggle": [Function],
            "visible": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Component>
                Lorem ipsum dolor sit amet
              </Component>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Lorem ipsum dolor sit amet",
              },
              "ref": null,
              "rendered": "Lorem ipsum dolor sit amet",
              "type": [Function],
            },
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "children": <Component>
              
            </Component>,
            "icon": <ForwardRef
              color="#fff"
              name="sort-down"
              type="FontAwesome"
            />,
            "primary": true,
            "title": "Second Element",
            "toggle": [Function],
            "visible": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "children": <Component>
              
            </Component>,
            "icon": <ForwardRef
              color="#fff"
              name="sort-down"
              type="FontAwesome"
            />,
            "primary": true,
            "title": "Third Element",
            "toggle": [Function],
            "visible": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
